shared/cleanup/languages/th.js

```diff
@@ -0,0 +1,25 @@
+export function sortSentences(sentences) {
+  return sentences.sort();
+}
+
+export function clean(sentences) {
+  return sentences.map((sentence) => {
+    return sentence
+      .replace(/[\u200b\u200c]/g, '')  // removes zero-width chars (occurs in some Thai texts)
+      .replace(/\?/g, ' ') // Thai does not use question mark
+      .replace(/!/g, ' ') // Thai does not use exclamation mark
+      .replace(/,/g, ' ') // Thai does not use comma
```

Comment:
Replacing `!` and `?` with a space can lead to these kind of sentences, is the space there correct? Additionally this would add a space at the end of a sentence if the mark was at the end. Can you also verify that the removal of these is indeed correct? There seem to be quite a few using question marks at the end as well as inside sentences.

ไหนบอกสิเจ็บตรงไหน?
ใครขโมยพายยัดไส้ผลไม้ไป?
เธอคือนักโทษของฉันคุณรู้ไหม! 'อัศวินแดงพูดในที่สุด -> this would then include two spaces
หรือไม่สามารถทำได้, เข้าใจมั้ย -> also would then include two spaces

Reply:
`.replace(/\s+/g, ' ')`
will condense multiple spaces to one

`.replace(/^\s+/, '')`
`.replace(/\s+$/, '')`
will remove spaces at the beginning and the end.

The reason I do it as a two-step process (replace those signs first with space, then condense any multiple spaces) is trying to preserve the delimiter semantic of those signs.

Consider modified example sentences here:
1) `ทำได้,เข้าใจ`
2) `ทำได้, เข้าใจ`

If you immediately remove comma (`replace(/,/g, '')`), we will get:
1a) `ทำได้เข้าใจ`
2a) `ทำได้ เข้าใจ`

If we do two-step process (`replace(/,/g, ' ')` then `replace(/\s+/g, ' ')`), we will get:
1b) `ทำได้ เข้าใจ`
2b) `ทำได้ เข้าใจ`

Reply:
So, to clarify, it's okay to replace a question/exclamation mark with a space and it's still semantically and grammatically correct? It mostly seems to be an issue inside sentences, as at the end it will just be trimmed (as you correctly noted and I completely missed).

```
คุณทราบเรื่องนี้จากที่ไหน?” ฉันถาม -> คุณทราบเรื่องนี้จากที่ไหน ” ฉันถาม (I have no idea where that quote comes from, but that's how the sentence is in the DB..)
'เปลี่ยนล็อบสเตอร์อีกแล้ว!' เขาตะโกนอย่างสุดเสียง -> 'เปลี่ยนล็อบสเตอร์อีกแล้ว ' เขาตะโกนอย่างสุดเสียง
```

Reply:
They will still be semantically and grammatically correct.

But for semantic, it is possible in some cases that some level of semantic can be lost.

According to Office of the Royal Society, exclamation mark (!) can has three usages [1]:

1) signal an interjection (e.g. oh !)
2) signal a word that mimics natural sound (in an instance that it is impossible to written down the exact sound with available characters) (e.g. kaboom !)
3) signal an order or a warning

For question mark (?), it has two usages:

1) signal a question
2) signal a doubt (e.g. "Tailand (Thailand?)")

[1] อัศเจรีย์ (exclamation mark) http://www.royin.go.th/?page_id=10433
[2] ปรัศนี (question mark) http://www.royin.go.th/?page_id=10418 

--

For the usage (1) of question mark, which is the most common, the same sentence still retain the same semantic without the mark.

'คุณทราบเรื่องนี้จากที่ไหน?' has the same meaning as 'คุณทราบเรื่องนี้จากที่ไหน' ('ที่ไหน' at the end of this sentence means 'where', and already signalling a form of question).

There no change in sound when people read out as well, when question mark is presented in Thai, so it should be more desirable to have only one of the two sentences in the database, instead of having both.

--

For exclamation mark, removing it may affect a bit the way people read out the sentence (prosody?). So it may be reasonable to have both 'again' and 'again!' in the database.

Anyway, if that the case, I think we need to make it clear in the guideline that if people see the exclamation mark they should read it differently - like more intense, louder, etc. depends on the context.

--

Thanks for asking more about this anyway. I have also learned from the Royal Society document that, in Thai, a space is needed before ! and ?. So I will add the rules for that, if in the end we decided to keep the sign.

Reply:
Btw, lots of sentences have unbalanced quotes. I think we should find way to clean that automatically.

Reply:
In that case, if I understand you correctly, I would suggest to not remove question and exclamation marks.

Reply:
Ok, then may be just normalized the spaces around the marks.

Reply:
@bact I agree with you. I think that removing question mark will not change context of the sentence in Thai language.

Reply:
I have updated the rules to be less aggressive.

They are now not removing either question mark or exclamation mark, only normalizing spaces around them (keep one space before and one space after the mark, per the Office of the Royal Society's guideline).

Reply:
@MichaelKohler would you mind to review this again please? We no longer remove question and exclamation marks, only add a space before and after them. Thank you.

Reply:
Will have a look tomorrow!